clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java

```diff
@@ -457,120 +539,224 @@ public void onResponse(final ClientResponse response) {
+                        new ArrayList<>(this.requestedPartitions),
+                        throwOnFetchStableOffsetUnsupported);
+            }
             return new NetworkClientDelegate.UnsentRequest(builder, coordinatorRequestManager.coordinator())
-                .whenComplete((r, t) -> onResponse(r.receivedTimeMs(), (OffsetFetchResponse) r.responseBody()));
+                    .whenComplete((r, t) -> onResponse(r.receivedTimeMs(), (OffsetFetchResponse) r.responseBody()));
         }
 
+        /**
+         * Handle request responses, including successful and failed.
+         */
         public void onResponse(
-            final long currentTimeMs,
-            final OffsetFetchResponse response) {
-            Errors responseError = response.groupLevelError(groupState.groupId);
+                final long currentTimeMs,
+                final OffsetFetchResponse response) {
+            Errors responseError = response.groupLevelError(groupId);
             if (responseError != Errors.NONE) {
                 onFailure(currentTimeMs, responseError);
                 return;
             }
             onSuccess(currentTimeMs, response);
         }
 
+        /**
+         * Handle failed responses. This will retry if the error is retriable, or complete the
+         * result future exceptionally in the case of non-recoverable or unexpected errors.
+         */
         private void onFailure(final long currentTimeMs,
                                final Errors responseError) {
             handleCoordinatorDisconnect(responseError.exception(), currentTimeMs);
             log.debug("Offset fetch failed: {}", responseError.message());
             if (responseError == COORDINATOR_LOAD_IN_PROGRESS) {
-                retry(currentTimeMs);
+                retry(currentTimeMs, responseError.exception());
+            } else if (responseError == Errors.STALE_MEMBER_EPOCH || responseError == Errors.UNKNOWN_MEMBER_ID) {
+                boolean retried = maybeRetryOnGroupError(currentTimeMs, responseError);
+                if (!retried) {
+                    log.error("OffsetFetch failed with {} and the consumer is not part " +
+                        "of the group anymore (it probably left the group, got fenced" +
+                        " or failed). The request cannot be retried and will fail.", responseError);
+                    future.completeExceptionally(responseError.exception());
+                } else {
+                    log.debug("OffsetFetch failed with {} but the consumer is still part" +
+                        " of the group, so the request will be retried with the latest " +
+                        "member ID and epoch.", responseError);
+                }
             } else if (responseError == Errors.NOT_COORDINATOR) {
                 // re-discover the coordinator and retry
                 coordinatorRequestManager.markCoordinatorUnknown("error response " + responseError.name(), currentTimeMs);
```

Comment:
It is a bit weird that we don't do this based on the same errors than for the offset commit.

Reply:
Totally agree that it's inconsistent, and I truly don't know why it's being handled differently in the legacy coordinator, or the implications that changing it would have. So should we keep it like this here just to ensure we handle it in the same way as the legacy coordinator does, and review it separately to better understand and then improve it? If agree I can file a jira for it.